File: labs/mathlogic/lab2/lab2_2.py
from typing import TypeVar, Generic, Optional
from lab2_1 import StructureNode

T = TypeVar('T')

class Queue(Generic[T]):
    def __init__(self):
        self.head: Optional[StructureNode[T]] = None
        self.tail: Optional[StructureNode[T]] = None
        self._size: int = 0

    def is_empty(self):
        return self._size == 0

    def size(self):
        return self._size

    def enqueue(self, data: T):
        node = StructureNode(data)
        if self.is_empty() == True:
            self.head = node
            self.tail = node
        elif self.is_empty() == False:
            self.tail.next = node
            self.tail = node
            
        self._size = self._size + 1

    def dequeue(self):
        if self.is_empty():
            raise Exception("Queue is empty!")
        
        data = self.head.data
        self.head = self.head.next
        self._size = self._size - 1
        
        if self.is_empty():
            self.tail = None
            
        return data

    def peek(self):
        if self.is_empty():
            raise Exception("Queue is empty!")
        return self.head.data

File: labs/mathlogic/lab2/lab2_2.ts
import { StructureNode } from "./lab2_1";

export class Queue<T> {
  head: StructureNode<T> | null = null;
  tail: StructureNode<T> | null = null;
  private _size: number = 0;

  is_empty() {
    return this._size == 0;
  }

  size() {
    return this._size;
  }

  enqueue(data: T) {
    const node = new StructureNode(data);
    switch (this.is_empty()) {
      case true:
        this.head = node;
        this.tail = node;
        break;
      case false:
        this.tail!.next = node;
        this.tail = node;
        break;
    }
    this._size = this._size + 1;
  }

  dequeue() {
    if (this.is_empty()) throw new Error("Queue is empty!");
    const data = this.head!.data;
    this.head = this.head!.next;
    this._size = this._size - 1;
    if (this.is_empty()) this.tail = null;
    return data;
  }

  peek() {
    if (this.is_empty()) throw new Error("Queue is empty!");
    return this.head!.data;
  }
}
